### ISSUE TITLE
Links within note inclusions are considered external.

### ISSUE THREAD

**Issue by Krrrrrrk** [obsidian-html/obsidian-html#178]
Love the work here. I apologize if I'm missing an option, but from my tests links to other notes within note inclusions are considered external, which prevents the .md files linked in the note inclusion being converted to .html unless Process All is enabled. This may break the path in the URL as well. 

For example:

NoteA:
![NoteB]

If NoteB has links to other notes, those don't get converted to .html

**Comment by dwrolvink** [comment]
Hi, thanks for trying the package and letting me know you're having an issue!

I can't replicate the error as of yet. 

The example code does not work for me, neither in Obsidian, nor does it get changed by ObsidianHtml. I've used `![[NoteB]]` for testing instead. This works as expected: it does not give me any link, just the contents of NoteB in NoteA.

There are a couple of quick checks that you can try:
- Are all the notes in the folder configured at `obsidian_folder_path_str` in the config yaml? (Any note outside that folder will be considered external by default)
- Does the link to the note work if you make it a normal link, or is it still an external link then?
- Does the name of the note contain '://` anywhere? 
- Can you change the filename to something containing only alphanumeric characters (like actually `NoteB`), does it work then?
- What does the output tell you when you append `-v` to the command that you run to generate the website? (e.g. `obsidianhtml -i config.yml -v`
 

If that leads nowhere, maybe you can share a bit more in depth information here:

- What is the version that you are using? (try `pip show obsidianhtml`)
- What are the contents of the config.yaml that you use?
- Can you share the verbose output?
- Can you create two test notes (one source linking to the second), test that they also have the issue, send their contents here and their filenames/paths?

Once I have a setup that contains the bug I can work on a way to correct it.

**Comment by dwrolvink** [comment]
Also curious what would happen if you add 
``` yaml
toggles:
    # Whether external http anchor links should have a target of "_blank"
  external_blank: False  
```

to your config.yml, if that would do anything
(but your inclusion being parsed as a link, instead of dumping the content is the real issue here I guess)

**Comment by dwrolvink** [comment]
Wow, sorry for the spam, but I just read your question entirely wrong. 

I just created the same setup:
```
# noteA
![[noteB]]
```
```
# noteB
[[noteC]]
```
```
# noteC
test
```

This works fine, in noteA I have the code of noteB and that links to noteC.html, no external links, it all works. This with the following config:

``` yaml
# The first folder that contains all obsidian files
obsidian_folder_path_str: 'ci/test_vault/'

# The note that will be used as the index.html 
obsidian_entrypoint_path_str: 'ci/test_vault/entrypoint.md'

# Input and output path of markdown files
md_folder_path_str:  'tmp/md'

# Markdown entrypoint path
md_entrypoint_path_str: 'tmp/md/index.md'

# Output path of HTML files
html_output_folder_path_str: 'tmp/html'

toggles:
    # Whether external http anchor links should have a target of "_blank"
  external_blank: False  
  # Opt-in/-out of Obsidian->Md conversion, set to False when using proper markdown as input
  compile_md: True

  # Opt-in/-out of Md->Html conversion, set to False when only wanting to get proper markdown from Obsidian
  compile_html: True

  # If this is false only the entrypoint and all its links (recursively) are processed
  # if true all the notes will be processed
  process_all: False

  # Can be overwritten ad-hoc by using "obsidianhtml -i config.yml -v" (the -v option)
  verbose_printout: False

  # This option should be False for Obsidian->Md, but can be True when compile_md == False
  # Setting it to True will cause an error when two files with the same file name are found anywhere in the input folder
  allow_duplicate_filenames_in_root: False

  # Sometimes linked images that should be local (based on the path) cannot be found.
  # Toggle this to False to disable warnings if that happens.
  warn_on_skipped_image: True

  # This will skip emptying output folders, if you want to implement this yourself
  no_clean: False

  # Whether the markdown interpreter assumes relative path when no / at the beginning of a link
  relative_path_md: True

  # Whether external http anchor links should have a target of "_blank"
  external_blank: False  
```

Could you share the actual links of the notes in Obsidian, and their respective actual filepaths in your vault, and in the markdown output folder?

Also it might be useful to know what version you are using. Could you share the output of `pip show obsidianhtml`?

**Comment by dwrolvink** [comment]
As far as I can get now with the reasoning: a link will be made external if the file in the link:
- is not of type `.md` 
- if the link includes `\\\\`, `://` or `:\\`

Included notes are simply added to the links, so that they are parsed in full as well (so that the links from those notes are also parsed). Really not sure what is going on. Cannot reproduce atm with the lastest version.

**Comment by dwrolvink** [comment]
Just to double check, what happens if you add the link to noteC in noteA as a normal link?

**Comment by Krrrrrrk** [comment]
> Just to double check, what happens if you add the link to noteC in noteA as a normal link?

That works correctly. I'm doing a test with a new vault and getting the other info for you now.

**Comment by Krrrrrrk** [comment]
![image](https://user-images.githubusercontent.com/34396790/157561741-68d91873-9f09-4074-b44b-9efdda7d3a48.png)

I copied your config for the test, changing the folder and entry point paths.
```
obsidian_folder_path_str: 'C:\vault\test' 
obsidian_entrypoint_path_str: 'C:\vault\test\start_here.md' 
```

Also tried:

```
obsidian_folder_path_str: 'test\' 
obsidian_entrypoint_path_str: 'test\start_here.md'
```


Has the same result. 

![image](https://user-images.githubusercontent.com/34396790/157561852-c769f605-20c7-4339-83a3-556406ae1a07.png)

Uploaded the test vault here: [vault.zip](https://github.com/obsidian-html/obsidian-html/files/8219131/vault.zip)

If you prefer I can type out the filepaths used, the markdown, and the links in Obsidian instead of browsing that .zip

Thank you for helping with this.

**Comment by Krrrrrrk** [comment]
Turned on verbose logging, you can see it skip the link.md which is the note that should be included into "start_here".

![image](https://user-images.githubusercontent.com/34396790/157563221-66f42bc6-8121-48f4-b46d-7a2083b92349.png)

**Comment by dwrolvink** [comment]
Yes, I managed to reproduce the error on windows (I normally develop on linux)
Seems to be an issue with file paths being split on `/` and windows paths not being converted to posix format in this specific case.

Working on it!

**Comment by dwrolvink** [comment]
Alright I think I have the fix for your testcase. Try the following:

( Using your test vault with entrypoint at `C:\vault\test\start_here.md` )

- In your config.yml:
  -  Set `relative_md: False` (is already present, but current value is True)
  -  Set all the [folder paths](https://obsidian-html.github.io/Configurations/Configuration%20Options.html#!paths) to absolute paths (e.g. `md_folder_path_str: 'C:\vault\tmp\md' )
- Get the newest version & run that:
  ``` bash
  git clone git@github.com:obsidian-html/obsidian-html.git
  cd obsidian-html
  # example run
  python -m obsidianhtml -i 'c:\vault\config.yml'
  # test
  python -m http.server --directory c:\vault\temp\html
  ```

If that works then I can update the pip package.

`relative_md` is something I had to add when I wanted to convert directly from a markdown project from gitlab. The relative nature caused issues there. But this is not really used much as far as I can tell, and it is poorly implemented/tested.

To get this to work in the included code *and* the orginal note, the relative nature would force us to write extra code to second-hand parse the code for the included note... 

But bottom line is: for obsidian-->html there should be nothing wrong with just absolute paths everywhere (that is, absolute relative to the vault dir, to make it simple :smiling_face_with_tear: )

tl;dr:
- you need to set toggles/relative_md to False
- you need to get the most recent version as the current package doesn't respect the above setting in all the necessary places
- to use the dev version, you need to call the module from the directory it is located in, so you need to update your paths to be absolute to output them where you did previously

Let me know if that works!

**Comment by dwrolvink** [comment]
For if anyone else reads this in the future:

This is a good minimal setup, because for this error to occur, the following should happen:
- noteB should be not on the same level as noteA (i.e. in a folder)
- noteC should be not on the same level as noteB, nor noteA

What happens is that the relative md path of noteC as calculated from noteB is different than what it should be if calculated from noteA. Because these are different, the code that is included from noteB should be different than the actual code of noteB.

Process_all: true fixes this because in the html conversion step, the relative links are replaced by absolute links and it is pretty smart with that, but the link was wrong in the md conversion step, could not be copied over to the md output, and thus the md note will not be present for the html conversion step.

Though process_all should not have to be enabled for this to work.

relative_md: False fixes this (starting with the current working code) because it forces all the urls to be in absolute format in relation to the webroot, this makes is so that the link to the target note will be the same, irregardless of the location of the source note.

#59

**Comment by Krrrrrrk** [comment]
> 

Awesome, this is working correctly now. No issues with the new package and the changes you advised. Thank you.

**Comment by dwrolvink** [comment]
Happy to hear that!

You can go back to the pip version now once you do `pip install obsidianhtml --upgrade` that should get you version 1.0.2 which has this fixed.

Once [this issue](https://github.com/obsidian-html/obsidian-html/issues/179) has been closed, it should be possible again to have `relative_path_md: True` again, might you (or any one reading this) need to have navigable md output :)

More information: https://obsidian-html.github.io/Instructions/How%20to%20resolve%20common%20issues.html#!links-in-included-notes-do-not-work-show-up-as-external

**Comment by dwrolvink** [comment]
Just a quick fyi @Krrrrrrk : the issue has been completely resolved in the latest commit of the master branch, this will also be included in version 1.0.3 when I get around to publishing it.

So if you need `relative_path_md: True` in the future, that is an option again from now on.
